File: regions/system/library/dart/session_client/lib/src/session_client.dart
import 'session.dart';

abstract class SessionClient {
  Future<Session> create(CreateSessionRequest req);
  Future<Session?> get(String id);
  Future<Session> refresh(RefreshSessionRequest req);
  Future<void> revoke(String id);
  Future<List<Session>> listUserSessions(String userId);
  Future<int> revokeAll(String userId);
}

class InMemorySessionClient implements SessionClient {
  final _sessions = <String, Session>{};
  int _counter = 0;

  @override
  Future<Session> create(CreateSessionRequest req) async {
    _counter++;
    final id = _counter.toString();
    final now = DateTime.now();
    final session = Session(
      id: id,
      userId: req.userId,
      token: 'tok-$id',
      expiresAt: now.add(Duration(seconds: req.ttlSeconds)),
      createdAt: now,
      metadata: req.metadata ?? {},
    );
    _sessions[id] = session;
    return session;
  }

  @override
  Future<Session?> get(String id) async {
    return _sessions[id];
  }

  @override
  Future<Session> refresh(RefreshSessionRequest req) async {
    final existing = _sessions[req.id];
    if (existing == null) {
      throw StateError('Session not found: ${req.id}');
    }
    final refreshed = existing.copyWith(
      expiresAt: DateTime.now().add(Duration(seconds: req.ttlSeconds)),
      token: 'tok-${req.id}-refreshed',
    );
    _sessions[req.id] = refreshed;
    return refreshed;
  }

  @override
  Future<void> revoke(String id) async {
    final existing = _sessions[id];
    if (existing == null) return;
    _sessions[id] = existing.copyWith(revoked: true);
  }

  @override
  Future<List<Session>> listUserSessions(String userId) async {
    return _sessions.values
        .where((s) => s.userId == userId)
        .toList();
  }

  @override
  Future<int> revokeAll(String userId) async {
    var count = 0;
    for (final entry in _sessions.entries.toList()) {
      if (entry.value.userId == userId && !entry.value.revoked) {
        _sessions[entry.key] = entry.value.copyWith(revoked: true);
        count++;
      }
    }
    return count;
  }
}

File: regions/system/library/swift/session-client/Sources/K1s0SessionClient/SessionClient.swift
import Foundation

public protocol SessionClient: Sendable {
    func create(req: CreateSessionRequest) async throws -> Session
    func get(id: String) async throws -> Session?
    func refresh(req: RefreshSessionRequest) async throws -> Session
    func revoke(id: String) async throws
    func listUserSessions(userId: String) async throws -> [Session]
    func revokeAll(userId: String) async throws -> Int
}

public actor InMemorySessionClient: SessionClient {
    private var sessions: [String: Session] = [:]

    public init() {}

    public func create(req: CreateSessionRequest) async throws -> Session {
        let now = Date()
        let session = Session(
            id: UUID().uuidString,
            userId: req.userId,
            token: UUID().uuidString,
            expiresAt: now.addingTimeInterval(Double(req.ttlSeconds)),
            createdAt: now,
            metadata: req.metadata
        )
        sessions[session.id] = session
        return session
    }

    public func get(id: String) async throws -> Session? {
        sessions[id]
    }

    public func refresh(req: RefreshSessionRequest) async throws -> Session {
        guard let existing = sessions[req.id] else {
            throw SessionError.sessionNotFound(id: req.id)
        }
        guard !existing.revoked else {
            throw SessionError.sessionRevoked(id: req.id)
        }
        let refreshed = Session(
            id: existing.id,
            userId: existing.userId,
            token: existing.token,
            expiresAt: Date().addingTimeInterval(Double(req.ttlSeconds)),
            createdAt: existing.createdAt,
            metadata: existing.metadata
        )
        sessions[req.id] = refreshed
        return refreshed
    }

    public func revoke(id: String) async throws {
        guard let existing = sessions[id] else {
            throw SessionError.sessionNotFound(id: id)
        }
        let revoked = Session(
            id: existing.id,
            userId: existing.userId,
            token: existing.token,
            expiresAt: existing.expiresAt,
            createdAt: existing.createdAt,
            revoked: true,
            metadata: existing.metadata
        )
        sessions[id] = revoked
    }

    public func listUserSessions(userId: String) async throws -> [Session] {
        sessions.values.filter { $0.userId == userId }
    }

    public func revokeAll(userId: String) async throws -> Int {
        let userSessions = sessions.values.filter { $0.userId == userId && !$0.revoked }
        for session in userSessions {
            let revoked = Session(
                id: session.id,
                userId: session.userId,
                token: session.token,
                expiresAt: session.expiresAt,
                createdAt: session.createdAt,
                revoked: true,
                metadata: session.metadata
            )
            sessions[session.id] = revoked
        }
        return userSessions.count
    }
}
